File: main/java/com/marsrovers/control/CommandExecutor.java
package com.marsrovers.control;


import com.marsrovers.units.Rover;

import java.util.List;
import java.util.stream.Collectors;

public class CommandExecutor {

    Rover execute(Rover rover, Command command) {

        if (!command.isExecutable()) {
            return rover;
        }

        Rover delegatedRover = rover;

        List<CommandDelegation> delegations = command
                                                .getMoveList()
                                                .stream()
                                                .map(CommandDelegation::fromString)
                                                .collect(Collectors.toList());


        for (CommandDelegation delegation : delegations) {
            delegatedRover = delegation.delegate(delegatedRover);
        }

        return delegatedRover;

    }

}

File: main/java/com/marsrovers/control/CommandExecutor.kt
package com.marsrovers.control


import com.marsrovers.units.Rover
import java.util.stream.Collectors

class CommandExecutor {

    fun execute(rover: Rover, command: Command): Rover {

        var delegatedRover = rover

        command
                .moveList
                .map { move -> CommandDelegation.fromString(move) }
                .forEach { delegation -> delegatedRover = delegation.delegate(delegatedRover) }

        return delegatedRover

    }

}
